### PR TITLE
AdamW β1/β2 sweep across all four datasets

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AdamW's momentum parameters β1 and β2 have never been ablated in this research programme. Every run so far uses the PyTorch defaults β1=0.9 and β2=0.999. These are sensible defaults for image classification and language modelling, but CFD mesh regression has a very different loss landscape: heavy spatial heterogeneity, asymmetric error scales across different regions, and cosine LR cycling that creates periodic gradient magnitude changes. In this setting:

- **β1 (first-moment / momentum):** Smaller β1 (e.g. 0.85) makes the optimiser more reactive to recent gradients, potentially beneficial when LR is changing rapidly in the cosine warm-up; larger β1 (e.g. 0.95) adds more momentum smoothing.
- **β2 (second-moment / RMS adaptation):** Smaller β2 (e.g. 0.99) makes the adaptive scale respond faster to gradient magnitude changes — helpful when different mesh regions have very different gradient scales. Larger β2 (e.g. 0.9999) slows adaptation, behaving more like SGD with momentum in terms of per-parameter scaling.

This sweep tests whether the β defaults are truly optimal for our tasks or if a different β pair closes meaningful error gaps.

## Current Baselines

| Dataset | Best val metric | Best PR |
|---------|----------------|---------|
| TandemFoil (TF) | 22.537 | #2924 (Lion — TF uses Lion, not AdamW) |
| TandemFoil Paper (TFP) | 0.00434 | #2979 |
| AirfRANS (AF) | 0.000482 | #2951 |
| DrivAerML (DM) | 3.997% | #2898 |

**Note:** TandemFoil uses Lion as its optimizer, so the β sweep only applies to TFP, AF, and DM. For TF, run the champion Lion config as a baseline validation run.

## Instructions

Use `--wandb_group adamw-beta-sweep` for all runs so they group in W&B.

### Step 0 — Check if β flags exist

```bash
cd target/icml2026 && python train.py --help | grep -i beta
```

If `--adamw-beta1` / `--adamw-beta2` (or similar) exist, use them directly. If not, find where `optim.AdamW(...)` is called in `train.py` and note the exact instantiation — then pass `betas=(β1, β2)` there for your runs, and document what you found in your results comment.

### Phase 1 — β2 sweep on AirfRANS (fastest feedback loop)

AF trains quickly; use it to identify the best β2 before committing GPU time on DM.

```bash
# β2=0.99 (faster second-moment adaptation)
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --adamw-beta1 0.9 --adamw-beta2 0.99 \
  --wandb_group adamw-beta-sweep

# β2=0.9999 (slower adaptation, closer to SGD-m scaling)
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --adamw-beta1 0.9 --adamw-beta2 0.9999 \
  --wandb_group adamw-beta-sweep
```

### Phase 2 — β1 sweep on AirfRANS (using best β2 from Phase 1)

```bash
# β1=0.85 — more reactive momentum
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --adamw-beta1 0.85 --adamw-beta2 <best-beta2-from-phase1> \
  --wandb_group adamw-beta-sweep

# β1=0.95 — more momentum smoothing
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --adamw-beta1 0.95 --adamw-beta2 <best-beta2-from-phase1> \
  --wandb_group adamw-beta-sweep
```

### Phase 3 — Best β pair on TFP and DM

Apply the winning (β1, β2) pair to TFP and DM.

**TandemFoil Paper (TFP):**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer adamw --lr 5e-4 --cosine-t-max 150 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --epochs 999 \
  --adamw-beta1 <best-beta1> --adamw-beta2 <best-beta2> \
  --wandb_group adamw-beta-sweep
```

**DrivAerML (DM):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --adamw-beta1 <best-beta1> --adamw-beta2 <best-beta2> \
  --wandb_group adamw-beta-sweep
```

**TandemFoil (TF):** TF uses Lion (not AdamW), so run the champion Lion config as a control/validation:
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --wandb_group adamw-beta-sweep
```

## Mandatory config reminders

- `--enable-fourier` REQUIRED for all datasets — do not omit
- TF: `--ema-decay 0.999` (Lion optimizer)
- TFP: `--no-use-ema` (per baseline #2979)
- AF: `--no-use-ema` (per baseline #2951)
- DM: `--no-use-ema` (mandatory) + `SENPAI_MAX_EPOCHS=9999`
- `--epochs 999` for all runs

## Success criteria

Beat the per-dataset baseline on any dataset. Target metrics:
- TFP < 0.00434
- AF < 0.000482
- DM < 3.997%

If β2=0.99 or β1=0.85 shaves >0.5% off DM or improves TFP/AF, that is a strong result worth merging. Report all W&B run IDs in your results comment.

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.

**Comment by tcapelle** [comment]
STUDENT jet:

## Results

### Summary

**No β variant beat any dataset baseline.** The PyTorch defaults (β1=0.9, β2=0.999) are near-optimal for our CFD tasks. All deviations either matched baseline within noise or caused training instability/divergence.

### Phase 1 — β2 sweep on AirfRANS (β1=0.9 fixed)

| β2 | Best Val (surface_mse) | Best Epoch | W&B Run | Status |
|------|----------------------|-----------|---------|--------|
| 0.99 | 0.000940 | 381 | [`akibwc2l`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/akibwc2l) | **Diverged** — grad_norm=Infinity, all batches clipping |
| **0.9999** | **0.000485** | 688 | [`fmsc3cth`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/fmsc3cth) | Stable, ~tied with baseline |
| 0.999 (baseline) | 0.000482 | — | — | — |

**Winner: β2=0.9999** (only stable option better than β2=0.99, but didn't beat baseline)

### Phase 2 — β1 sweep on AirfRANS (β2=0.9999 fixed)

| β1 | Best Val (surface_mse) | Best Epoch | W&B Run | Status |
|------|----------------------|-----------|---------|--------|
| 0.85 | 0.000632 | 534 | [`vjgz2d5a`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/vjgz2d5a) | Stable but 31% worse |
| 0.95 | 0.000763 | 291 | [`xlq6pgab`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/xlq6pgab) | **Diverged to NaN** |
| 0.9 (baseline) | 0.000482 | — | — | — |

**β1=0.9 remains optimal.** Both alternatives degraded performance significantly.

### Phase 3 — Best β pair (β1=0.9, β2=0.9999) on TFP and DM + TF Lion control

| Dataset | Config | Best Val | Baseline | W&B Run | Status |
|---------|--------|----------|----------|---------|--------|
| TFP | β1=0.9, β2=0.9999 | NaN (epoch 1) | 0.00434 | [`2f03ifjq`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/2f03ifjq) | **Diverged to NaN immediately** |
| DM | β1=0.9, β2=0.9999 | 26.21% (epoch 22) | 3.997% | [`xj13t1vr`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/xj13t1vr) | **Catastrophic** — 6.5× worse, grad_norm ~1e12 |
| TF | Lion control (no β change) | 22.77 (epoch 311) | 22.537 | [`kmk4m8e3`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/kmk4m8e3) | Stable, close to baseline |

### Commands used

```bash
# Phase 1 — β2 sweep
CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --adamw-beta1 0.9 --adamw-beta2 0.99 --agent jet --wandb-name "jet/af-beta2-0.99" --wandb-group adamw-beta-sweep
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --adamw-beta1 0.9 --adamw-beta2 0.9999 --agent jet --wandb-name "jet/af-beta2-0.9999" --wandb-group adamw-beta-sweep

# Phase 2 — β1 sweep (with best β2=0.9999)
CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --adamw-beta1 0.85 --adamw-beta2 0.9999 --agent jet --wandb-name "jet/af-beta1-0.85-beta2-0.9999" --wandb-group adamw-beta-sweep
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --adamw-beta1 0.95 --adamw-beta2 0.9999 --agent jet --wandb-name "jet/af-beta1-0.95-beta2-0.9999" --wandb-group adamw-beta-sweep

# Phase 3 — Cross-dataset
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil_paper --optimizer adamw --lr 5e-4 --cosine-t-max 150 --no-use-ema --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --epochs 999 --adamw-beta1 0.9 --adamw-beta2 0.9999 --agent jet --wandb-name "jet/tfp-beta1-0.9-beta2-0.9999" --wandb-group adamw-beta-sweep
SENPAI_MAX_EPOCHS=9999 CUDA_VISIBLE_DEVICES=1 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --adamw-beta1 0.9 --adamw-beta2 0.9999 --agent jet --wandb-name "jet/dm-beta1-0.9-beta2-0.9999" --wandb-group adamw-beta-sweep
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 --agent jet --wandb-name "jet/tf-lion-control" --wandb-group adamw-beta-sweep
```

### Peak memory
- AirfRANS runs: ~17-32 GB VRAM
- TFP: ~30 GB VRAM
- DM: ~6 GB VRAM
- TF Lion: ~48 GB VRAM

### What happened

**The hypothesis did not work.** The PyTorch default AdamW betas (β1=0.9, β2=0.999) are near-optimal for all our CFD datasets:

1. **β2=0.99 (faster adaptation)** caused gradient explosion on AirfRANS — the faster second-moment response amplified gradient magnitude oscillations instead of smoothing them.
2. **β2=0.9999 (slower adaptation)** was essentially tied with baseline on AirfRANS (0.000485 vs 0.000482) but **catastrophically failed** on TFP (NaN from epoch 1) and DM (26% vs 4% baseline, with gradient norms ~1e12). The slower adaptation appears to prevent the optimizer from properly scaling per-parameter learning rates in these more complex loss landscapes.
3. **β1 variations** (0.85 and 0.95 with β2=0.9999) both degraded AirfRANS performance. β1=0.95 diverged to NaN; β1=0.85 was stable but 31% worse.
4. **TF Lion control** confirmed the Lion baseline is reproducible (22.77 vs 22.537 baseline).

The key insight: CFD mesh regression with cosine LR cycling and gradient clipping is well-served by the standard β defaults. The spatial heterogeneity and asymmetric error scales that motivated this hypothesis are apparently already handled by the combination of gradient clipping (gc=1.0) and the default adaptive scaling.

### Suggested follow-ups
- The near-tie between β2=0.999 and β2=0.9999 on AirfRANS suggests the second-moment time constant isn't a major sensitivity — the optimizer landscape for AF is well-conditioned at this LR/architecture.
- The catastrophic failure of β2=0.9999 on TFP and DM (but not AF) suggests these datasets have sharper curvature or more variable gradient scales that need faster second-moment adaptation. This could be investigated further by per-dataset β2 tuning, but the expected gains are small given the AF result.
- Weight decay and gradient clipping interactions may be more impactful levers than betas for these tasks.

### Code changes
Added `--adamw-beta1` and `--adamw-beta2` CLI flags to `train.py` with defaults matching PyTorch (0.9, 0.999). These are wired into the `build_optimizer` function and logged to W&B via the config dict.
